lib/Dialect/Torch/Transforms/RefineTypes.cpp

```diff
@@ -587,24 +598,39 @@ getPromotedResultTypeAssumingNonZeroRank(MLIRContext *context,
                                /*skipRankCheck=*/true);
 }
 
-// Get the MLIR type of the tensor dtype given the dtype integer value and the
-// input dtype. When DType is None the type is inferred from the input dtype.
-static void fillInDTypeGivenDTypeIntAndInputDType(ValueKnowledge &knowledge,
-                                                  Value dtype,
-                                                  Type inputDType) {
+void TypeAnalyzer::fillInDTypeGivenDTypeIntAndInputDType(
+    ValueKnowledge &knowledge, Value dtype, Type inputDType) {
   assert(isBuiltInType(inputDType) && "`inputDType` must be a builtin type");
   int64_t dtypeInt;
   if (dtype.getType().isa<Torch::NoneType>())
     knowledge.dtype = inputDType;
   else if (matchPattern(dtype, m_TorchConstantInt(&dtypeInt)))
     knowledge.dtype = getTypeForDTypeInteger(dtype.getContext(), dtypeInt);
+  else if (auto primDtypeOp = dyn_cast<PrimDtypeOp>(dtype.getDefiningOp())) {
+    auto element = getLatticeElement(primDtypeOp.a());
+    assert(!element.isUninitialized() && "no refined type for source");
+
+    switch (element.getValue().kind) {
+    case torch_upstream::TypeKind::TensorType:
+      knowledge.dtype = element.getValue().dtype;
+      break;
+
+    case torch_upstream::TypeKind::NumberType:
```

Comment:
do we need to handle this case? I don't think it is legal to call prim.dtype on a scalar value.

Reply:
You're right, `prim.dtype` should never have a scalar operand.  I'll remove the case.

Reply:
Removed.